spark/src/main/scala/io/substrait/spark/expression/ToWindowFunction.scala

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.substrait.spark.expression
+
+import io.substrait.spark.expression.ToWindowFunction.{fromSpark, fromSparkFollowing, fromSparkPreceding}
+
+import org.apache.spark.sql.catalyst.expressions.{CurrentRow, Expression, FrameType, Literal, OffsetWindowFunction, RangeFrame, RowFrame, SpecifiedWindowFrame, UnboundedFollowing, UnboundedPreceding, UnspecifiedFrame, WindowExpression, WindowSpecDefinition}
+import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression
+import org.apache.spark.sql.types.{IntegerType, LongType}
+
+import io.substrait.`type`.Type
+import io.substrait.expression.{Expression => SExpression, ExpressionCreator, FunctionArg, WindowBound}
+import io.substrait.expression.Expression.WindowBoundsType
+import io.substrait.expression.WindowBound.{CURRENT_ROW, Following, Preceding, UNBOUNDED}
+import io.substrait.extension.SimpleExtension
+import io.substrait.relation.ConsistentPartitionWindow.WindowRelFunctionInvocation
+
+import scala.collection.JavaConverters
+
+abstract class ToWindowFunction(functions: Seq[SimpleExtension.WindowFunctionVariant])
+  extends FunctionConverter[SimpleExtension.WindowFunctionVariant, WindowRelFunctionInvocation](
+    functions) {
+
+  override def generateBinding(
+      sparkExp: Expression,
+      function: SimpleExtension.WindowFunctionVariant,
+      arguments: Seq[FunctionArg],
+      outputType: Type): WindowRelFunctionInvocation = {
+
+    val (frameType, lower, upper) = sparkExp match {
+      case WindowExpression(_: OffsetWindowFunction, _) =>
+        (WindowBoundsType.ROWS, UNBOUNDED, CURRENT_ROW)
```

Comment:
What is an `OffsetWindowFunction`. Does it somehow override the frame bounds?

Reply:
That's my understanding. TBH I'm not sure exactly how it should translate to substrait, but it's ignored by spark (it matches the lag/lead functions)
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala#L404

Reply:
You actually should get the frame here as well properly. Otherwise the bound will be incorrect for other consumers (unbounded preceeding and current row will not work for Lead, for example) and Spark will throw. You'll see that if you add `logicalPlan2.show()` into `SubstraitPlanTestBase::assertSqlToSubstraitRoundTrip` and re-run the lag/lead test. (We should also make the tests always evaluate the converted plans to ensure they are actually valid, I'll try to get to that)

Can be fixed by just removing this `case`.

Reply:
Ok, I've removed it for now